### PR TITLE
Draft of ES declarative config

### PR DESCRIPTION
**Pull request by anyasabo**
Fix https://github.com/elastic/cloud-on-k8s/issues/3598

Opening up a draft to get feedback on the approach, still needs a lot of shoring up to do. But it "works". 

## An overview

- users provide us a list of URLs and an optional JSON body that URL should contain, and the elasticsearch resource it should apply to
- we do a get against that URL
  - if it's a 200 and the response body is a superset of the user-specified JSON body, we consider this operation successful and move on to the next operation in the list
  - if it is a 404 or a 200 but the response is not a superset of the user-specified body, we send a put to that URL with the user-specified body

## Example manual test

```yaml
---
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.9.1
  nodeSets:
  - name: default
    count: 1
    config:
      node.master: true
      node.data: true
      node.ingest: true
      node.store.allow_mmap: false
---
apiVersion: elasticsearchconfig.k8s.elastic.co/v1alpha1
kind: ElasticsearchConfig
metadata:
  name: quickstart-config
spec:
  operations:
  - url: "/_component_template/component_template1"
    body: |-
      {
        "template": {
          "mappings": {
            "properties": {
              "@timestamp": {
                "type": "date"
              }
            }
          }
        }
      }
  - url: "/_component_template/other_component_template"
    body: |-
      {
        "template": {
          "mappings": {
            "properties": {
              "ip_address": {
                "type": "ip"
              }
            }
          }
        }
      }
  elasticsearchRef:
    name: quickstart
```

```
PASSWORD=$(kubectl get secret quickstart-es-elastic-user -o go-template='{{.data.elastic | base64decode}}')

kubectl port-forward service/quickstart-es-http 9200

 curl -s -u "elastic:$PASSWORD" -k "https://localhost:9200/_component_template/component_template1" | jq

 curl -s -u "elastic:$PASSWORD" -k "https://localhost:9200/_component_template/other_component_template" | jq
```

## Open questions

- does the association interface make sense to use here? it has a meaning in other resources because the operator is configuring a resource (e.g. kibana) to create an association. here, the operator is making the calls itself, so it has a slightly different meaning. I cargo culted it in because I was used to it, but thinking about it more I'm not sure it's necessary

- ES sometimes supports additional arguments in PUTs that it does not support for GETs (e.g. [`verify=false`](https://www.elastic.co/guide/en/elasticsearch/reference/current/snapshots-register-repository.html#snapshots-repository-verification) for snapshot repo creation), but we provide no mechanism for that here. It's not clear to me how important that might be though.

- does checking specifically for 404s/200s make sense? I'm not sure if there's other status codes ES use that make sense for us to proceed on

- now is a good time for naming/structure feedback too

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -171,6 +171,7 @@ go-run:
 			-tags "$(GO_TAGS)" \
 			./cmd/main.go manager \
 				--development \
+				--enable-leader-election=false \
```
Comment (anyasabo) [inline Makefile:174]: This is just to enable quicker iteration when running locally and could be moved to a separate PR

**Comment by sebgl** [comment]
I added a few comments about the general approach in the issue: https://github.com/elastic/cloud-on-k8s/issues/3598#issuecomment-702783723

File: pkg/controller/esconfig/esconfig_controller.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	getResp, _ := client.Request(ctx, get)
+
+	// nothing exists at this url yet, time to create it
+	if getResp.StatusCode == http.StatusNotFound {
+		logger.V(1).Info("resource does not exist yet", "url", opURL)
+		return true, nil
+	}
+
+	// TODO checking for a 200 might be wrong, really any status code not 200 might mean we need to update. but a 200 indicates we can and should compare the bodies
+	// TODO should bodies always be required to be specified? I think probably? I don't know of any ES API that you can send an empty PUT. maybe we require at a minimum "{}"?
```
Comment (pebrc) [inline pkg/controller/esconfig/esconfig_controller.go:253]: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-data-stream.html

**Comment by anyasabo** [comment]
I'll go ahead and close this out while we investigate an alternative approach (a single fire and forget job) that I'll detail in the issue
